### ISSUE TITLE
Interactive Abstract Pattern Generation Javascript Demo

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#1349]
Interactive Abstract Pattern Generation Javascript Demo<br>
<br>
https://ift.tt/1rt1sgN<br>
<br>
<br>
<br>

<div>
<section><p>
</p>
<center>
<a target="_blank" href="http://otoro.net/ml/netart/">
<img src="http://blog.otoro.net/assets/20160424/interactive_neural_art.large.png">
</a>
</center>
<br>
<center>
Interactive Javascript <a href="http://otoro.net/ml/netart/">Demo</a> for Abstract Pattern Generation.
<br>
<code>
<br>
<a target="_blank" href="https://github.com/hardmaru/netart-js">GitHub</a>
</code>
</center>

<p>After a recent <a href="http://blog.otoro.net/2016/03/25/generating-abstract-patterns-with-tensorflow/">blog post</a> where I demonstrated a really simple method to generate high resolution abstract images with neural networks and latent vectors, many people were wondering if they can use this method without the need to setup TensorFlow.  Although there were some <a href="http://blog.otoro.net/2015/06/19/neural-network-generative-art/">code</a> available previously in Javascript, it wasn’t general enough to use as a tool for a digital artist.</p>

<p>So I took the <a href="http://blog.otoro.net/2015/06/19/neural-network-generative-art/">Javascript code</a> previously written and spent an hour or two to fine tuned it into a simple <a href="http://otoro.net/ml/netart/">web app</a>.  Karpathy’s <a href="https://github.com/karpathy/recurrentjs">recurrent.js</a> library makes it really easy to implement highly customised neural networks in JS, and adopts a computational graph type of method similar to modern libraries.  Using these tools, and JQuery, I incorporated latent vectors into the previous JS model, as well as options for different colour models, and alpha-transparency modes.  In addition, the user is able to specify the size and depth of the generator neural network.</p>

<table>
<tr>
<td>
<img src="http://blog.otoro.net/assets/20160401/svg/generator.svg">
<br>
<center>
<i>Architecture of the Generator Network</i>
</center>
</td>
</tr>
</table>

<p>The generator network we have implemented takes in a latent vector <code>Z</code> of 2 elements only for this demo.  The network is just a stack of fully-connected <code>tanh</code> layers, like before.  The depth and size of the network, and also the image resolution of the output can all be customised in the web app.</p>

<h3 id="usage">Usage</h3>

<p>The best way to use this web app is to start with a low resolution for faster processing, of say around <code>160px</code>.  You can play around with different architecture configurations by hitting the <code>restart</code> button, which will randomise all of the weights of the neural network and display the updated picture with the new settings.</p>

<center>
<img src="http://blog.otoro.net/assets/20160424/interactive_neural_art.hsl.png">
<br>
</center>

<p>Different colour modes are now supported.  I have implemented Black and White, Red Green Blue, CMYK, HSV, and HSL colour models, so basically the output sigmoid layer of the neural network will correspond to one of these colour model layers.  In addition, I have added the option to have an extra alpha-transparancy layer if we want to use an extra output to fill in the alpha channel.</p>

<p>After you are satisfied with the network architecture, you can also try to explore the latent space of the network by adjusting the latent vector <code>Z1</code> and <code>Z2</code> and hitting the <code>redraw</code> button to see how the image changes.  After you arrive at an image you like, if the resolution is too small during the trial and error phase, you can also increase the resolution, and hit the <code>redraw</code> button to render the same image in larger size.  The <code>save</code> button will save the canvase into a <code>.png</code> file.</p>

<p>Try the web app demo <a href="http://otoro.net/ml/netart/">here</a>.</p>

<h3 id="various-examples">Various Examples</h3>
<center>

</center>

<h3 id="future-work">Future Work</h3>

<p>The implementation is currently not optimised, meaning that I query each pixel.  It will be much faster to try to batch process the entire image, or even each line, rather than each pixel at a time.  I have done such optimisations with <a href="http://blog.otoro.net/2015/07/31/neurogram/">Neurogram</a>, but that was a much more intense project, rather than just an hour hack.  If we try use WebGL to optimise matrix operations, we maybe able to perform near-realtime rendering in the future.</p>

<p>It should also be possible to export certain simple types of TensorFlow trained networks into JSON and get <code>recurrent.js</code> to run those networks, given the somewhat similar computational graph nature of both libraries.</p>

</section>

    
  
  </div><br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via blog.otoro.net http://blog.otoro.net<br>
<br>
October 31, 2018 at 03:39PM
